### PR TITLE
Add `Method Not Allowed` in the Router

### PR DESCRIPTION
**Pull request by isgj**
When the same path is registered but with another method respond with `405` instead of `404`

### CODE REVIEW

File: src/router.rs [hunk 1]
```diff
@@ -61,6 +61,11 @@ impl<State: 'static> Router<State> {
             // if not then fallback to the behavior of HTTP GET else proceed as usual
 
             self.route(path, http::Method::GET)
+        } else if self.method_map.values().any(|r| r.recognize(path).is_ok()) {
+            Selection {
+                endpoint: &method_not_allawed,
```
Comment (yoshuawuyts) [inline src/router.rs]:
```suggestion
                endpoint: &method_not_allowed,
```

File: src/router.rs [hunk 2]
```diff
@@ -73,3 +78,7 @@ impl<State: 'static> Router<State> {
 fn not_found_endpoint<State>(_cx: Request<State>) -> BoxFuture<'static, Response> {
     Box::pin(async move { Response::new(http::StatusCode::NOT_FOUND.as_u16()) })
 }
+
+fn method_not_allawed<State>(_cx: Request<State>) -> BoxFuture<'static, Response> {
```
Comment (yoshuawuyts) [inline src/router.rs]:
```suggestion
fn method_not_allowed<State>(_cx: Request<State>) -> BoxFuture<'static, Response> {
```

File: src/router.rs [hunk 3]
```diff
@@ -61,6 +61,11 @@ impl<State: 'static> Router<State> {
             // if not then fallback to the behavior of HTTP GET else proceed as usual
 
             self.route(path, http::Method::GET)
+        } else if self.method_map.values().any(|r| r.recognize(path).is_ok()) {
```
Comment (yoshuawuyts) [inline src/router.rs]: This code contains quite a few branches which can be hard to follow; could you maybe add a comment here explaining what this does?

**Review by yoshuawuyts — APPROVED** [review]
This patch is really good; all for better diagnostics!

Left a few nits, should be good to merge when addressed. Thanks so much for this!

**Comment by isgj** [comment]
Happy to hear that.
